### ISSUE TITLE
semantic activation maps generation

### ISSUE THREAD

**Issue by so-ux** [hustvl/MaskAdapter#16]
![Image](https://github.com/user-attachments/assets/b60d97c9-0108-4023-8d64-6911193f49f3)
Can the semantic activation maps generation code be made public? I'm not sure how you obtained the maps using the mask adapter. From the description in the article, it seems more like you used pooled features to generate the maps. 

In addition, the feature with 32 channels constructed in multiple places in the code, I am not sure if 32 has any special meaning because there is no formula derivation

**Comment by owl-10** [comment]
For semantic activation maps generation code , you can refer to [this](https://github.com/hustvl/MaskAdapter/blob/fd1b74392d6e939108fc3ec473c611a1a2914cc6/mask_adapter/mask_adapter.py#L357) and [arch of mask-adapter](https://github.com/hustvl/MaskAdapter/blob/main/mask_adapter/modeling/meta_arch/mask_adapter_head.py). Mask-Adapter receive masks and clip feature , output 16 semantic activation maps,you can refer to [this](https://github.com/hustvl/MaskAdapter/blob/fd1b74392d6e939108fc3ec473c611a1a2914cc6/mask_adapter/modeling/meta_arch/mask_adapter_head.py#L108). I don't know where the feature with 32 channels constructed, can you show me where it is.

**Comment by so-ux** [comment]
![Image](https://github.com/user-attachments/assets/fcac753a-a158-409c-846c-78c6d28ae03c)

![Image](https://github.com/user-attachments/assets/5146b3e4-0684-4b9c-92f0-36dcd1b2371c)
Thank you for your reply. This is an excellent job. 

What I mean by 32 is that setting max_matches=self. num_pred_masks to 8 in the image can exactly make the second dimension of mask_pred 32. Is this some kind of fixed requirement? 

I am currently doing 12 classes classification. Are there any rules for setting up self. num_pred_masks? 

Also, I noticed that you used a trained mask adapter as a plugin to combine with Sam2. Have you tried using Sam2 for Mask Generator during training? I think if Sam2 doesn't predict masks at the beginning, it may not rely on the mask adapter to get the correct results in the future.

**Comment by owl-10** [comment]
Thank you for your thoughtful question!

During the second training stage, in order to help the adapter adapt to predicted masks, we mix both ground-truth masks and predicted masks for training. Since predicted masks do not have class labels, we use an IoU-based matcher to assign labels: for each ground-truth mask, we find the predicted mask with the highest IoU and assign the class label of the ground-truth mask to it.

The number of predicted masks (self.num_pred_masks) mainly depends on the number of ground-truth masks in an image. If your GPU memory is sufficient, you can increase the max_matches accordingly. In our case, we used 24 ground-truth masks and 8 predicted masks, making up a total of 32 masks. Among them, 8 predicted masks are matched with ground-truth masks.

In our experiments on large-scale datasets like COCO-Stuff and COCO-Panoptic (with over 100 categories), setting num_pred_masks = 8 was sufficient.

Regarding your question about using SAM2 as the mask generator during training: we directly use the pre-trained mask adapter as a plugin with SAM2. However, it's also a good idea to include SAM2 during training so that the adapter can learn to classify masks generated by SAM2. You're welcome to explore this direction.
